nightfall/api.py

```diff
@@ -1,153 +1,262 @@
-                    'conditionSetUUID': self.condition_set
-                }
-            }
-
-            response = requests.post(
-                url=self._url,
-                headers=self._headers,
-                data=json.dumps(data)
-            )
 
-            # Logs for Debugging
-            self.logger.debug(f"HTTP Request URL: {response.request.url}")
-            self.logger.debug(f"HTTP Request Body: {response.request.body}")
-            self.logger.debug(
-                f"HTTP Request Headers: {response.request.headers}")
+        if not policy_uuid and not detection_rule_uuids and not detection_rules:
+            raise NightfallUserError("at least one of policy_uuid, detection_rule_uuids or detection_rules required",
+                                     40001)
+
+        response = self._file_scan_initialize(location)
+        _validate_response(response, 200)
+        result = response.json()
+        session_id, chunk_size = result['id'], result['chunkSize']
+
+        uploaded = self._file_scan_upload(session_id, location, chunk_size)
+        if not uploaded:
+            raise NightfallSystemError("File upload failed", 50000)
 
-            self.logger.debug(f"HTTP Status Code: {response.status_code}")
-            self.logger.debug(f"HTTP Response Headers: {response.headers}")
-            self.logger.debug(f"HTTP Response Text: {response.text}")
+        response = self._file_scan_finalize(session_id)
+        _validate_response(response, 200)
 
-            response.raise_for_status()
-            findings = response.json()
+        response = self._file_scan_scan(session_id,
+                                        detection_rules=detection_rules,
+                                        detection_rule_uuids=detection_rule_uuids,
+                                        webhook_url=webhook_url, policy_uuid=policy_uuid)
+        _validate_response(response, 200)
+        parsed_response = response.json()
 
-            for idx, d in enumerate(chunk):
-                if findings[idx] is not None:
-                    all_findings[d] = findings[idx]
-                else:
-                    all_findings[d] = None
+        return parsed_response["id"], parsed_response["message"]
+
+    def _file_scan_initialize(self, location: str):
+        data = {
+            "fileSizeBytes": os.path.getsize(location)
+        }
+        response = requests.post(
+            url=self.FILE_SCAN_INITIALIZE_ENDPOINT,
+            headers=self._headers,
+            data=json.dumps(data)
+        )
+
+        return response
+
+    def _file_scan_upload(self, session_id: str, location: str, chunk_size: int):
+
+        def read_chunks(fp, chunk_size):
+            ix = 0
+            while True:
+                data = fp.read(chunk_size)
+                if not data:
+                    break
+                yield ix, data
+                ix = ix + 1
+
+        def upload_chunk(id, data, headers):
```

Comment:
nit: I'm not really sure why this would need to be it's own nested function, it seems like it would actually reduce the total lines and increase readability to just inline the logic below